src/main/java/net/glowstone/util/pattern/PortalShape.java

```diff
@@ -0,0 +1,248 @@
+            in.getY() + (face.getModY() * offset),
+            in.getZ() + (face.getModZ() * offset)
+        );
+    }
+
+    /**
+     * Get the distance to one edge of the possible portal. Returns -1 if no end is reached.
+     *
+     * @param location the location from where to check. Not modified in process.
+     * @param face     the side of the edge to check for
+     * @return the distance or -1 in case of an invalid shape
+     */
+    private int getDistanceUntilEdge(Location location, BlockFace face) {
+        // Clone the location as Location objects are mutable
+        Location destLoc = location.clone();
+
+        int i;
+        for (i = 0; i < 22; ++i) {
+            // Move onwards
+            destLoc.add(face.getModX(), face.getModY(), face.getModZ());
+            // Break if either an obstacle or the end of the obsidian "line" is reached
+            if (!canBuildThrough(destLoc.getBlock().getType())
+                || downImmutable(destLoc).getBlock().getType() != Material.OBSIDIAN) {
+                break;
+            }
+        }
+        // If the target block is not obsidian, the portal is invalid.
+        return destLoc.getBlock().getType() == Material.OBSIDIAN ? i : -1;
+    }
+
+    /**
+     * Calculate the portals height.
+     *
+     * @return the height or 0 in case of failure
+     */
+    private int calculatePortalHeight() {
+        for (height = 0; height < 21; ++height) {
+            boolean flag = false;
+            for (int i = 0; i < width; ++i) {
+                Location location = offsetImmutable(bottomLeft, left.getOppositeFace(), i)
+                    .add(0, height, 0);
+                Material material = location.getBlock().getType();
+
+                if (!canBuildThrough(material)) {
+                    flag = true;
+                    break;
+                }
+
+                if (material == Material.PORTAL) {
+                    ++portalBlockCount;
+                }
+
+                if (i == 0) {
+                    material = offsetImmutable(location, left, 1).getBlock().getType();
+                    if (material != Material.OBSIDIAN) {
+                        flag = true;
+                        break;
+                    }
+                } else if (i == width - 1) {
+                    material = offsetImmutable(location, left, -1).getBlock().getType();
+                    if (material != Material.OBSIDIAN) {
+                        flag = true;
+                        break;
+                    }
+                }
+            }
+            if (flag) {
+                break;
+            }
+        }
+
+        for (int j = 0; j < width; ++j) {
+            if (offsetImmutable(bottomLeft, left, -j).add(0, height, 0).getBlock().getType()
+                != Material.OBSIDIAN) {
+                height = 0;
+                break;
+            }
+        }
+
+        if (height <= 21 && height >= 3) {
+            return height;
+        } else {
```

Comment:
The else statement is unnecessary because you are returning in the if-statement.